src/Checkbox.js

```diff
@@ -0,0 +1,195 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import styled from 'react-emotion';
+import { rem } from 'polished';
+
+import Box from './Box';
+import Flex from './Flex';
+import Text from './Text';
+import Icon from './Icon';
+
+const CheckIcon = ({ fill = 'currentColor', ...props }) => (
+  <svg width="10" height="7" viewBox="0 0 10 7" {...props}>
+    <path
+      fill={fill}
+      fillRule="evenodd"
+      d="M4 4.586L1.707 2.293A1 1 0 1 0 .293 3.707l3 3a.997.997 0 0 0 1.414 0l5-5A1 1 0 1 0 8.293.293L4 4.586z"
+    />
+  </svg>
+);
+
+CheckIcon.propTypes = {
+  fill: PropTypes.string.isRequired
+};
+
+const MinusIcon = ({ fill = 'currentColor', ...props }) => (
+  <svg width="16" height="16" viewBox="0 0 16 16" {...props}>
+    <path
+      fill={fill}
+      fillRule="evenodd"
+      d="M11 7H5c-.55 0-1 .45-1 1s.45 1 1 1h6c.55 0 1-.45 1-1s-.45-1-1-1z"
+    />
+  </svg>
+);
+
+MinusIcon.propTypes = {
+  fill: PropTypes.string.isRequired
+};
+
+const Label = styled(Flex.withComponent('label'))`
+  cursor: ${props => (props.disabled ? 'not-allowed' : 'pointer')};
+`;
+
+const Input = styled(Box.withComponent('input'))`
+  clip: rect(1px, 1px, 1px, 1px);
+  height: 1px;
+  position: absolute;
+  white-space: nowrap;
+  width: 1px;
+  opacity: 0;
+  overflow: hidden;
+  padding: 0;
+`;
+
+const StyledCheckbox = ({ children, ...props }) => (
+  <Flex
+    bg="white"
+    border="1px solid"
+    borderColor="grey60"
+    borderRadius="small"
+    color="white"
+    alignItems="center"
+    justifyContent="center"
+    width={rem('16px')}
+    height={rem('16px')}
+    {...props}
+  >
+    {children}
+  </Flex>
+);
+
+const Checkbox = ({
+  id,
+  name,
+  label,
+  disabled,
+  isInvalid,
+  checked,
+  onChange,
+  value,
+  indeterminate,
+  icon,
+  ...props
+}) => (
+  <Label alignItems="center" my="regular" {...props}>
+    <Input
+      id={id}
+      type="checkbox"
+      name={name}
+      value={value}
+      checked={checked || indeterminate}
+      onChange={onChange}
+      disabled={disabled}
+      aria-invalid={isInvalid}
+    />
+    {!disabled && checked ? (
+      <StyledCheckbox
+        borderColor="blue"
+        bg={indeterminate ? 'white' : 'blue'}
+        color={indeterminate ? 'blue' : 'white'}
+      >
+        {indeterminate ? <MinusIcon /> : <CheckIcon />}
+      </StyledCheckbox>
+    ) : (
+      !disabled && <StyledCheckbox />
+    )}
+    {disabled && (
+      <StyledCheckbox
+        borderColor="grey20"
+        bg={checked ? 'grey20' : 'white'}
+        color={checked ? 'grey50' : 'white'}
+      >
+        {indeterminate ? <MinusIcon /> : <CheckIcon />}
+      </StyledCheckbox>
```

Comment:
Instead of duplicating the calls to `StyledCheckbox` like this, could we have done this via emotion's styling changes based off props? https://emotion.sh/docs/styled

Reply:
We could have. To me, this feels more human readable and less redundant than repeating `${props => props.theme...}` again and again. This also maps directly to how I'd write my design specs, and how I imagine product folks communicating, which is a huge win in my book because everyone's speaking the same language and reading out of the same book.

One thing I wish I could figure out is how to get the icon logic into `StyledCheckbox` - I couldn't get it to render all the states properly.

Reply:
I tend to disagree for a couple of reasons.

1. We’re deviating from an established pattern.
2. Shared language is important but doesn’t need to apply to an implementation detail.

The main issue here is that it’s easy for a change to creep up that needs to be applied into both places where this is duplicated in the button. Something that can easily be overlooked.

Reply:
I would argue that a design system is both a shared language _and_ implementation details. The two need to be aligned.

I'm also not sure I follow the change creep issue - all shared styles are defined in <StyledCheckbox />, so anything that needs to be applied in both places would be applied there.